libs/execution_context/ebpf_maps.c

```diff
@@ -43,33 +43,125 @@ typedef struct _ebpf_core_object_map
+// Fewer partitions will result in more contention on the lock, but more partitions will consume more memory.
+#define EBPF_LRU_MAXIMUM_PARTITIONS 8
+
 /**
- * @brief Each LRU entry tracks a key for an entry stored in a LRU map. The entry is stored in one of two lists: hot or
- * cold. The hot list is for items that have been accessed in the current generation, where a generation is defined as a
- * period of time during which max_entries/EBPF_LRU_GENERATION_COUNT elements have been accessed in the map. The cold
- * list is for items that have not been accessed in the current generation. When the hot list reaches
+ * @brief The BPF_MAP_TYPE_LRU_HASH is a hash table that stores a limited number of entries. When the map is full, the
+ * least recently used entry is removed to make room for a new entry. The map is implemented as a hash table with a pair
+ * of double-linked lists per partition. The hot list contains entries that have been accessed in the current
+ * generation, and the cold list contains entries that have not been accessed in the current generation. When entries in
+ * the cold list are accessed, they are moved to the hot list. When the hot list reaches
  * max_entries/EBPF_LRU_GENERATION_COUNT, the hot list is merged into the cold list, a new generation is started, and
- * the hot list is cleared.  When space is needed in the map, the cold list is trimmed to make room for new entries. The
- * cold list is trimmed by removing the oldest entries in the list, which are part of the oldest generation. The benefit
- * is that the lock only needs to be acquired when moving items between generations, and not on every access.
+ * the hot list is cleared. When space is needed an entry is selected from a cold list and is removed from the
+ * hash table.
+ *
+ * key history is stored along with the value in the map. The hash table then provides callbacks to the map to update
+ * the key history when an entry is accessed, updated, or deleted.
+ *
+ * key history can be in multiple partitions, with different generation and last-used-time values. To determine
+ * the actual last used time of a key, the map must iterate over all the partitions and find the maximum last-used-time.
+ *
+ * The link in the ebpf_lru_entry_t with the highest corresponding last-used-time is the actual last used time of the
+ * key. All other links in that ebpf_lru_entry_t are considered stale and are ignored.
+ *
+ * Key history is stored in a separate partition from the value and is returned as an opaque blob of memory.
+ *
+ * ebpf_lru_entry_partition_t is an untyped block of memory containing the following:
+ * struct ebpf_lru_entry_t {
+ *  ebpf_list_entry_t list_entry[partition_count];
+ *  size_t generation[partition_count];
+ *  size_t last_used_time[partition_count];
+ *  uint8_t key[key_size];
+ * };
+ *
+ * This structure is not expressible in C, so we use macros to calculate the offsets of the fields in the structure and
```

Comment:
Why not group all fields in the same partition together, as in:

```
struct ebpf_lru_entry_partition_t {
    ebpf_list_entry_t list_entry;
    size_t generation;
    size_t last_used_time;
    uint8_t key;
};
struct ebpf_lru_entry_t {
    struct ebpf_lru_entry_partition_t partition[partition_count];
};
```

Reply:
Two issues:
1) We can't have 2 variable sized fields in a C struct (both the partitions and the key are variable sized).
2) Computing the effective timestamp for a key is run once per partition, so I chose to make sure the last_used_time are all on the same cache line.